File: web/src/generators/ContourMapGenerator.js
import { Generator } from '../core/Generator';
import { ParameterDefinition } from '../core/ParameterDefinition';
import { SvgCanvas } from '../core/SvgCanvas';
import { SeededRandom } from '../utils/SeededRandom';
import { createNoise2D } from 'simplex-noise';
import alea from 'alea';

export class ContourMapGenerator extends Generator {
    getId() {
        return "contour-map";
    }

    getDisplayName() {
        return "Contour Map";
    }

    getParameterDefinitions() {
        return [
            ParameterDefinition.selection("Preset", "Custom", ["Custom", "Mountain Range", "Rolling Hills", "Island Archipelago", "Canyon Ridges"], "Select a predefined style"),
            ParameterDefinition.integer("Contour Lines", 25, 5, 60, "Number of elevation contour lines"),
            ParameterDefinition.doubleVal("Scale", 3.0, 0.5, 10.0, "Noise scale (higher = more zoomed in features)"),
            ParameterDefinition.integer("Octaves", 4, 1, 8, "Noise octaves for terrain complexity"),
            ParameterDefinition.doubleVal("Persistence", 0.5, 0.1, 0.9, "How much each octave contributes"),
            ParameterDefinition.integer("Resolution", 250, 50, 400, "Grid resolution for contour extraction"),
            ParameterDefinition.bool("Major Lines", true, "Draw thicker lines every 5th contour"),
            ParameterDefinition.integer("Seed", 42, 1, 9999, "Random seed for terrain generation"),
            ParameterDefinition.integer("Colors", 1, 1, 6, "Number of plotter layers")
        ];
    }

    onParameterChanged(paramName, newValue, currentValues) {
        if (paramName === "Preset" && typeof newValue === "string") {
            switch (newValue) {
                case "Mountain Range":
                    currentValues["Contour Lines"] = 30;
                    currentValues["Scale"] = 3.0;
                    currentValues["Octaves"] = 5;
                    currentValues["Persistence"] = 0.5;
                    currentValues["Resolution"] = 300;
                    currentValues["Major Lines"] = true;
                    currentValues["Colors"] = 1;
                    return true;
                case "Rolling Hills":
                    currentValues["Contour Lines"] = 15;
                    currentValues["Scale"] = 2.0;
                    currentValues["Octaves"] = 3;
                    currentValues["Persistence"] = 0.6;
                    currentValues["Resolution"] = 250;
                    currentValues["Major Lines"] = true;
                    currentValues["Colors"] = 1;
                    return true;
                case "Island Archipelago":
                    currentValues["Contour Lines"] = 20;
                    currentValues["Scale"] = 4.0;
                    currentValues["Octaves"] = 4;
                    currentValues["Persistence"] = 0.45;
                    currentValues["Resolution"] = 300;
                    currentValues["Major Lines"] = true;
                    currentValues["Colors"] = 2;
                    return true;
                case "Canyon Ridges":
                    currentValues["Contour Lines"] = 40;
                    currentValues["Scale"] = 5.0;
                    currentValues["Octaves"] = 6;
                    currentValues["Persistence"] = 0.55;
                    currentValues["Resolution"] = 350;
                    currentValues["Major Lines"] = true;
                    currentValues["Colors"] = 1;
                    return true;
                default:
                    return false;
            }
        }
        if (paramName !== "Preset" && currentValues["Preset"] !== "Custom") {
            currentValues["Preset"] = "Custom";
            return true;
        }
        return false;
    }

    generate(params) {
        const numContours = params["Contour Lines"] || 25;
        const scale = params["Scale"] || 3.0;
        const octaves = params["Octaves"] || 4;
        const persistence = params["Persistence"] || 0.5;
        const resolution = params["Resolution"] || 250;
        const majorLines = params["Major Lines"] !== undefined ? params["Major Lines"] : true;
        const seed = params["Seed"] || 42;
        const numColors = params["Colors"] || 1;

        const width = params["width"] || 1000;
        const height = params["height"] || 1000;

        const canvas = new SvgCanvas(width, height, numColors);

        // Create noise function
        const prng = alea(seed);
        const noise2D = createNoise2D(prng);

        // Build terrain height field using fractal noise
        const cellW = width / resolution;
        const cellH = height / resolution;
        const field = [];
        let minVal = Infinity;
        let maxVal = -Infinity;

        for (let iy = 0; iy <= resolution; iy++) {
            field[iy] = [];
            for (let ix = 0; ix <= resolution; ix++) {
                const nx = (ix / resolution) * scale;
                const ny = (iy / resolution) * scale;

                let val = 0;
                let amp = 1;
                let freq = 1;
                let maxAmp = 0;

                for (let o = 0; o < octaves; o++) {
                    val += amp * noise2D(nx * freq, ny * freq);
                    maxAmp += amp;
                    amp *= persistence;
                    freq *= 2;
                }

                val /= maxAmp; // Normalize to [-1, 1]
                field[iy][ix] = val;

                if (val < minVal) minVal = val;
                if (val > maxVal) maxVal = val;
            }
        }

        // Extract contour lines
        const defaultStrokeWidth = canvas.strokeWidth;

        for (let c = 0; c < numContours; c++) {
            const t = c / (numContours - 1);
            const level = minVal + t * (maxVal - minVal);
            const isMajor = majorLines && (c % 5 === 0);
            const layerIndex = c % numColors;

            if (isMajor) {
                canvas.setStrokeWidth(defaultStrokeWidth * 2.0);
            } else {
                canvas.setStrokeWidth(defaultStrokeWidth * 0.7);
            }

            const segments = this.marchingSquares(field, resolution, resolution, cellW, cellH, level);
            const paths = this.connectSegments(segments);

            for (const path of paths) {
                if (path.length < 2) continue;
                // Simplify path with Douglas-Peucker
                const simplified = this.simplifyPath(path, cellW * 0.3);
                let d = `M ${simplified[0].x.toFixed(2)} ${simplified[0].y.toFixed(2)}`;
                for (let i = 1; i < simplified.length; i++) {
                    d += ` L ${simplified[i].x.toFixed(2)} ${simplified[i].y.toFixed(2)}`;
                }

                // Use raw SVG with specific stroke-width for major/minor distinction
                if (isMajor) {
                    canvas.addRaw(layerIndex, `<path d='${d}' stroke-width='${(defaultStrokeWidth * 2.0).toFixed(2)}' />`);
                } else {
                    canvas.addRaw(layerIndex, `<path d='${d}' stroke-width='${(defaultStrokeWidth * 0.7).toFixed(2)}' />`);
                }
            }
        }

        canvas.setStrokeWidth(defaultStrokeWidth);
        return canvas.toSvg();
    }

    simplifyPath(path, tolerance) {
        if (path.length <= 2) return path;

        let maxDist = 0;
        let maxIndex = 0;

        const first = path[0];
        const last = path[path.length - 1];

        for (let i = 1; i < path.length - 1; i++) {
            const d = this.pointLineDistance(path[i], first, last);
            if (d > maxDist) {
                maxDist = d;
                maxIndex = i;
            }
        }

        if (maxDist > tolerance) {
            const left = this.simplifyPath(path.slice(0, maxIndex + 1), tolerance);
            const right = this.simplifyPath(path.slice(maxIndex), tolerance);
            return left.slice(0, -1).concat(right);
        }

        return [first, last];
    }

    pointLineDistance(point, lineStart, lineEnd) {
        const dx = lineEnd.x - lineStart.x;
        const dy = lineEnd.y - lineStart.y;
        const lenSq = dx * dx + dy * dy;
        if (lenSq === 0) return Math.sqrt((point.x - lineStart.x) ** 2 + (point.y - lineStart.y) ** 2);

        const t = Math.max(0, Math.min(1, ((point.x - lineStart.x) * dx + (point.y - lineStart.y) * dy) / lenSq));
        const projX = lineStart.x + t * dx;
        const projY = lineStart.y + t * dy;
        return Math.sqrt((point.x - projX) ** 2 + (point.y - projY) ** 2);
    }

    marchingSquares(field, nx, ny, cellW, cellH, level) {
        const segments = [];

        for (let iy = 0; iy < ny; iy++) {
            for (let ix = 0; ix < nx; ix++) {
                const v0 = field[iy][ix];
                const v1 = field[iy][ix + 1];
                const v2 = field[iy + 1][ix + 1];
                const v3 = field[iy + 1][ix];

                const b0 = v0 >= level ? 1 : 0;
                const b1 = v1 >= level ? 1 : 0;
                const b2 = v2 >= level ? 1 : 0;
                const b3 = v3 >= level ? 1 : 0;

                const caseIndex = b0 | (b1 << 1) | (b2 << 2) | (b3 << 3);
                if (caseIndex === 0 || caseIndex === 15) continue;

                const x0 = ix * cellW;
                const y0 = iy * cellH;
                const x1 = (ix + 1) * cellW;
                const y1 = (iy + 1) * cellH;

                const interp = (va, vb, a, b) => {
                    const t = (level - va) / (vb - va);
                    return a + t * (b - a);
                };

                const top = { x: interp(v0, v1, x0, x1), y: y0 };
                const right = { x: x1, y: interp(v1, v2, y0, y1) };
                const bottom = { x: interp(v3, v2, x0, x1), y: y1 };
                const left = { x: x0, y: interp(v0, v3, y0, y1) };

                const addSeg = (a, b) => segments.push({ a, b });

                switch (caseIndex) {
                    case 1: addSeg(top, left); break;
                    case 2: addSeg(top, right); break;
                    case 3: addSeg(left, right); break;
                    case 4: addSeg(right, bottom); break;
                    case 5: addSeg(top, right); addSeg(left, bottom); break;
                    case 6: addSeg(top, bottom); break;
                    case 7: addSeg(left, bottom); break;
                    case 8: addSeg(left, bottom); break;
                    case 9: addSeg(top, bottom); break;
                    case 10: addSeg(top, left); addSeg(right, bottom); break;
                    case 11: addSeg(right, bottom); break;
                    case 12: addSeg(left, right); break;
                    case 13: addSeg(top, right); break;
                    case 14: addSeg(top, left); break;
                }
            }
        }

        return segments;
    }

    connectSegments(segments) {
        if (segments.length === 0) return [];

        const paths = [];
        const used = new Array(segments.length).fill(false);
        const EPS = 0.5;

        const near = (p1, p2) => Math.abs(p1.x - p2.x) < EPS && Math.abs(p1.y - p2.y) < EPS;

        for (let i = 0; i < segments.length; i++) {
            if (used[i]) continue;
            used[i] = true;

            const path = [segments[i].a, segments[i].b];
            let changed = true;

            while (changed) {
                changed = false;
                for (let j = 0; j < segments.length; j++) {
                    if (used[j]) continue;
                    const seg = segments[j];
                    const last = path[path.length - 1];
                    const first = path[0];

                    if (near(last, seg.a)) {
                        path.push(seg.b); used[j] = true; changed = true;
                    } else if (near(last, seg.b)) {
                        path.push(seg.a); used[j] = true; changed = true;
                    } else if (near(first, seg.a)) {
                        path.unshift(seg.b); used[j] = true; changed = true;
                    } else if (near(first, seg.b)) {
                        path.unshift(seg.a); used[j] = true; changed = true;
                    }
                }
            }

            paths.push(path);
        }

        return paths;
    }
}

File: src/main/java/org/trostheide/generativeArt/ContourMapGenerator.java
package org.trostheide.generativeArt;

import org.trostheide.generativeArt.core.ArtGenerator;
import org.trostheide.generativeArt.core.ParameterDefinition;
import org.trostheide.generativeArt.core.SvgCanvas;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class ContourMapGenerator implements ArtGenerator {

    @Override
    public String getId() {
        return "contour-map";
    }

    @Override
    public String getDisplayName() {
        return "Contour Map";
    }

    @Override
    public List<ParameterDefinition> getParameterDefinitions() {
        return List.of(
                ParameterDefinition.selection("Preset", "Custom",
                        Arrays.asList("Custom", "Mountain Range", "Rolling Hills", "Island Archipelago", "Canyon Ridges"),
                        "Select a predefined style"),
                ParameterDefinition.integer("Contour Lines", 25, 5, 60, "Number of elevation contour lines"),
                ParameterDefinition.doubleVal("Scale", 3.0, 0.5, 10.0,
                        "Noise scale (higher = more zoomed in features)"),
                ParameterDefinition.integer("Octaves", 4, 1, 8, "Noise octaves for terrain complexity"),
                ParameterDefinition.doubleVal("Persistence", 0.5, 0.1, 0.9, "How much each octave contributes"),
                ParameterDefinition.integer("Resolution", 250, 50, 400, "Grid resolution for contour extraction"),
                ParameterDefinition.bool("Major Lines", true, "Draw thicker lines every 5th contour"),
                ParameterDefinition.integer("Seed", 42, 1, 9999, "Random seed for terrain generation"),
                ParameterDefinition.integer("Colors", 1, 1, 6, "Number of plotter layers"));
    }

    @Override
    public boolean onParameterChanged(String paramName, Object newValue, Map<String, Object> currentValues) {
        if ("Preset".equals(paramName) && newValue instanceof String) {
            switch ((String) newValue) {
                case "Mountain Range":
                    currentValues.put("Contour Lines", 30);
                    currentValues.put("Scale", 3.0);
                    currentValues.put("Octaves", 5);
                    currentValues.put("Persistence", 0.5);
                    currentValues.put("Resolution", 300);
                    currentValues.put("Major Lines", true);
                    currentValues.put("Colors", 1);
                    return true;
                case "Rolling Hills":
                    currentValues.put("Contour Lines", 15);
                    currentValues.put("Scale", 2.0);
                    currentValues.put("Octaves", 3);
                    currentValues.put("Persistence", 0.6);
                    currentValues.put("Resolution", 250);
                    currentValues.put("Major Lines", true);
                    currentValues.put("Colors", 1);
                    return true;
                case "Island Archipelago":
                    currentValues.put("Contour Lines", 20);
                    currentValues.put("Scale", 4.0);
                    currentValues.put("Octaves", 4);
                    currentValues.put("Persistence", 0.45);
                    currentValues.put("Resolution", 300);
                    currentValues.put("Major Lines", true);
                    currentValues.put("Colors", 2);
                    return true;
                case "Canyon Ridges":
                    currentValues.put("Contour Lines", 40);
                    currentValues.put("Scale", 5.0);
                    currentValues.put("Octaves", 6);
                    currentValues.put("Persistence", 0.55);
                    currentValues.put("Resolution", 350);
                    currentValues.put("Major Lines", true);
                    currentValues.put("Colors", 1);
                    return true;
                default:
                    return false;
            }
        }
        if (!"Preset".equals(paramName) && !"Custom".equals(currentValues.get("Preset"))) {
            currentValues.put("Preset", "Custom");
            return true;
        }
        return false;
    }

    @Override
    public String generate(Map<String, Object> params) {
        int numContours = (int) params.getOrDefault("Contour Lines", 25);
        double scale = (double) params.getOrDefault("Scale", 3.0);
        int octaves = (int) params.getOrDefault("Octaves", 4);
        double persistence = (double) params.getOrDefault("Persistence", 0.5);
        int resolution = (int) params.getOrDefault("Resolution", 250);
        boolean majorLines = (boolean) params.getOrDefault("Major Lines", true);
        int seed = (int) params.getOrDefault("Seed", 42);
        int numColors = (int) params.getOrDefault("Colors", 1);

        double width = 1000, height = 1000;
        if (params.containsKey("width")) width = ((Number) params.get("width")).doubleValue();
        if (params.containsKey("height")) height = ((Number) params.get("height")).doubleValue();

        SvgCanvas canvas = new SvgCanvas(width, height, numColors);

        // Simple value noise using random grid
        Random rand = new Random(seed);
        int noiseGridSize = 64;
        double[][] noiseGrid = new double[noiseGridSize + 1][noiseGridSize + 1];
        for (int y = 0; y <= noiseGridSize; y++) {
            for (int x = 0; x <= noiseGridSize; x++) {
                noiseGrid[y][x] = rand.nextDouble() * 2 - 1;
            }
        }

        double cellW = width / resolution;
        double cellH = height / resolution;
        double[][] field = new double[resolution + 1][resolution + 1];
        double minVal = Double.MAX_VALUE, maxVal = -Double.MAX_VALUE;

        for (int iy = 0; iy <= resolution; iy++) {
            for (int ix = 0; ix <= resolution; ix++) {
                double nx = ((double) ix / resolution) * scale;
                double ny = ((double) iy / resolution) * scale;

                double val = 0, amp = 1, freq = 1, maxAmp = 0;
                for (int o = 0; o < octaves; o++) {
                    val += amp * sampleNoise(noiseGrid, noiseGridSize, nx * freq, ny * freq);
                    maxAmp += amp;
                    amp *= persistence;
                    freq *= 2;
                }
                val /= maxAmp;
                field[iy][ix] = val;
                if (val < minVal) minVal = val;
                if (val > maxVal) maxVal = val;
            }
        }

        double defaultStrokeWidth = 1.0;

        for (int c = 0; c < numContours; c++) {
            double t = (double) c / (numContours - 1);
            double level = minVal + t * (maxVal - minVal);
            boolean isMajor = majorLines && (c % 5 == 0);
            int layerIndex = c % numColors;

            List<double[]> segments = marchingSquares(field, resolution, cellW, cellH, level);
            String sw = isMajor
                    ? String.format(java.util.Locale.US, "stroke-width='%.2f'", defaultStrokeWidth * 2.0)
                    : String.format(java.util.Locale.US, "stroke-width='%.2f'", defaultStrokeWidth * 0.7);

            for (int i = 0; i + 1 < segments.size(); i += 2) {
                canvas.addRaw(layerIndex, String.format(java.util.Locale.US,
                        "<line x1='%.2f' y1='%.2f' x2='%.2f' y2='%.2f' %s />",
                        segments.get(i)[0], segments.get(i)[1],
                        segments.get(i + 1)[0], segments.get(i + 1)[1], sw));
            }
        }

        return canvas.toSvg();
    }

    private double sampleNoise(double[][] grid, int size, double x, double y) {
        // Wrap coordinates
        double wx = ((x % size) + size) % size;
        double wy = ((y % size) + size) % size;
        int ix = (int) wx, iy = (int) wy;
        double fx = wx - ix, fy = wy - iy;
        // Smoothstep
        fx = fx * fx * (3 - 2 * fx);
        fy = fy * fy * (3 - 2 * fy);

        int ix1 = (ix + 1) % size, iy1 = (iy + 1) % size;
        double v00 = grid[iy][ix], v10 = grid[iy][ix1];
        double v01 = grid[iy1][ix], v11 = grid[iy1][ix1];

        double top = v00 + fx * (v10 - v00);
        double bot = v01 + fx * (v11 - v01);
        return top + fy * (bot - top);
    }

    private List<double[]> marchingSquares(double[][] field, int res, double cellW, double cellH, double level) {
        List<double[]> segments = new ArrayList<>();
        for (int iy = 0; iy < res; iy++) {
            for (int ix = 0; ix < res; ix++) {
                double v0 = field[iy][ix], v1 = field[iy][ix + 1];
                double v2 = field[iy + 1][ix + 1], v3 = field[iy + 1][ix];
                int ci = (v0 >= level ? 1 : 0) | (v1 >= level ? 2 : 0) | (v2 >= level ? 4 : 0) | (v3 >= level ? 8 : 0);
                if (ci == 0 || ci == 15) continue;

                double x0 = ix * cellW, y0 = iy * cellH, x1 = (ix + 1) * cellW, y1 = (iy + 1) * cellH;
                double[] top = {interp(v0, v1, x0, x1, level), y0};
                double[] right = {x1, interp(v1, v2, y0, y1, level)};
                double[] bottom = {interp(v3, v2, x0, x1, level), y1};
                double[] left = {x0, interp(v0, v3, y0, y1, level)};

                switch (ci) {
                    case 1: segments.add(top); segments.add(left); break;
                    case 2: segments.add(top); segments.add(right); break;
                    case 3: segments.add(left); segments.add(right); break;
                    case 4: segments.add(right); segments.add(bottom); break;
                    case 5: segments.add(top); segments.add(right); segments.add(left); segments.add(bottom); break;
                    case 6: segments.add(top); segments.add(bottom); break;
                    case 7: segments.add(left); segments.add(bottom); break;
                    case 8: segments.add(left); segments.add(bottom); break;
                    case 9: segments.add(top); segments.add(bottom); break;
                    case 10: segments.add(top); segments.add(left); segments.add(right); segments.add(bottom); break;
                    case 11: segments.add(right); segments.add(bottom); break;
                    case 12: segments.add(left); segments.add(right); break;
                    case 13: segments.add(top); segments.add(right); break;
                    case 14: segments.add(top); segments.add(left); break;
                }
            }
        }
        return segments;
    }

    private double interp(double va, double vb, double a, double b, double level) {
        double t = (level - va) / (vb - va);
        return a + t * (b - a);
    }
}
